### DISCUSSION TITLE
Discussion: Regarding Contributing to the Project

### DISCUSSION THREAD

**Discussion by pizazzo** [PureDarwin/PureDarwin#104] [General]
Hey! I came across this project a few years ago but, I was wondering where I could get started communicating with other developers, and seeing where there are gaps to fill? 

Particularly, I was wondering about these: 
- How many of Apple's binary drivers that PureDarwin relies on has been reimplemented or have become missing over the years as Apple gradually stopped distributing them? Is there some sort of roadmap to completion in terms of binary drivers that potential collaborators have for reference?
   
   _**Update:** Providing a self-answer, a list of required binary drivers can be found here on the [Binary Drivers](https://github.com/PureDarwin/PureDarwin/wiki/BinaryDrivers) page on the wiki. It's not a pretty set of pages but, here's the progress on projects in PureDarwin, built upon [Darwin 10](https://github.com/PureDarwin/PureDarwin/wiki/Darwin_10) and [9](https://github.com/PureDarwin/PureDarwin/wiki/Darwin_9). Tables are pretty unreadable, but it tracks all the projects and how they're building currently. Current blockers can be found [here](https://github.com/PureDarwin/PureDarwin/wiki/Current_Blockers), and on issues as well._

- Do we have some reference on how to build the whole PureDarwin stack? Do we need an x86_64 macOS machine to do so, or are there possible alternatives? I currently have one of the new arm64 M1 Macs (and multiple x86_64 Linux machines, but that's besides the point).

   _**Update:** If the [homepage](http://www.puredarwin.org/) is still updated, we're at XNU 16.7(10.12.6). Build guide is [here](https://github.com/PureDarwin/Building-XNU-16.7-Kernel-Guide/). For unsupported platforms, it's not possible until an open-source alternative for `iig` from the [DriverKit SDK](https://developer.apple.com/documentation/driverkit/creating_a_driver_using_the_driverkit_sdk) is implemented. In terms of how to begin doing so, someone could fork [Darwin on ARM](https://github.com/darwin-on-arm) on [darwin-sdk](https://github.com/darwin-on-arm/darwin-sdk) and implement `iig`? It is a non-trivial task, however._

- Is there a particular place where we can communicate? The way to access current forms of communication is currently not really obvious or down. The forums have been down for a bit and I'm not too sure how to access the Freenode IRC channel.

I'm sorry if these are very obvious questions and if I sound really pushy, I'm just very interested in the development of the project and I was hoping to ask questions that me and other prospective future collaborators may want to ask to pick up the pace of development.

**Comment by nekename** [comment]
I actually came across this project a few days ago, and it looks really cool, but I can't find any of the ISOs for PureDarwin Xmas. 

As to your question about communication, they have a subreddit (although with no more than 10 posts) (here)[https://www.reddit.com/r/PureDarwin/].

The latest commit seems to be really recent, yet the project is still focused on Darwin 9.x and 10.x, with the latest version being 21.x, so I was wondering what the development state of this project is, is it active?

(also, the wiki is a mess, I know you moved from Google Sites recently, but there's HTML everywhere in Markdown pages, I've fixed a few, but I can't do 137 unfortunately)

Edit: I got the Xmas files, but it crashes (see my reddit post)

**Comment by pizazzo** [comment]
This reply has two purposes; just following up on the thread in general, and replying with things I found while digging around. This doesn't answer everything but I hope to point people to the path I've taken as I find things.

> How many of Apple's binary drivers that PureDarwin relies on has been reimplemented or has become missing over the years as Apple gradually stopped distributing them? Is there some sort of roadmap to completion in terms of binary drivers that potential collaborators have for reference?

Providing a self-answer, a list of required binary drivers can be found here on the [Binary Drivers](https://github.com/PureDarwin/PureDarwin/wiki/BinaryDrivers) page on the wiki. It's not a pretty set of pages but, here's the progress on projects in PureDarwin, built upon [Darwin 10](https://github.com/PureDarwin/PureDarwin/wiki/Darwin_10) and [9](https://github.com/PureDarwin/PureDarwin/wiki/Darwin_9). Tables are pretty unreadable, but it tracks all the projects and how they're building currently. Current blockers can be found [here](https://github.com/PureDarwin/PureDarwin/wiki/Current_Blockers), and on issues as well.

A little unclear is which version of Darwin/XNU is being used as the basis of the project and whether swapping between versions of Darwin is possible. There are no tags or branches describing changes in the Darwin version, so I couldn't quite pinpoint where to begin doing this (I looked for `Xmas` and `nano` branches and tags). However, we see Darwin 6-10 mentioned in the documentation; hopefully, we could see how accessing each version of Darwin would work? I think this is pretty important, especially since a lot of CPU and system support got removed with the new `AppleACPIPlatform` Platform Expert that replaces the Darwin 8 `AppleI386GenericPlatform` Platform Expert.

_**Update:** If the [homepage](http://www.puredarwin.org/) is still updated, we're at XNU 16.7(10.12.6). Build guide is [here](https://github.com/PureDarwin/Building-XNU-16.7-Kernel-Guide/)._

> Do we have some reference on how to build the whole PureDarwin stack? Do we need an x86_64 macOS machine to do so, or are there possible alternatives? I currently have one of the new arm64 M1 Macs (and multiple x86_64 Linux machines, but that's beside the point).

I can't really verify this step - mostly because I'm on an ARM machine so I don't only need to look for something that can run on ARM but also a toolchain that can cross-compile into x86_64/i386. But you can get started by reading the [DarwinBuild page](https://github.com/PureDarwin/PureDarwin/wiki/DarwinBuild) for reference on building bits of the OpenDarwin project and XNU. 

For my situation, I'm looking at using Darwin on Arm's [darwin-sdk](https://github.com/darwin-on-arm/darwin-sdk) that comes with the cc-toolchain and a bunch of other tools. It can cross-compile from `x86, x86_64, ARM` to `armv6, armv7, armv7s, arm64, i386, x86_64, x86_64h, armv6m, armv7k, armv7m, armv7em`. For those looking to build from another operating system (#61 and #66), the SDK supports Linux and Windows (via Cygwin) as well. 

I'm not too sure if it can act as a drop-in replacement for the DarwinBuild toolchain, or replace Xcode requirements on DarwinBuild to allow it to run on Linux (#57); I'll still need to test this out. This isn't a 'must' since my situation is certainly an outlier, but it could help expand the number of developers that can work on the project.

_**Update:** I found out on #82 that `iig` - a part of the [DriverKit SDK](https://developer.apple.com/documentation/driverkit/creating_a_driver_using_the_driverkit_sdk) which compiles `*.iig` IOKit interface generator files - is required by PureDarwin to compile several files, and it's closed-source. A quick skim over the [darwin-sdk](https://github.com/darwin-on-arm/darwin-sdk) by the [Darwin on Arm](https://github.com/darwin-on-arm/) people show that they don't really have it either. It's a non-trivial task, and will probably mean that a lot of the driver work will still have to be built on a macOS host at the very least. Darwin on Arm's XNU version is much lower than PD's, so their SDK doesn't have an `iig` command to compile these user-space drivers._

It's not very clear from the wiki but, here is a [folder in the repo](https://github.com/PureDarwin/PureDarwin/tree/master/setup) covering how one would package the built binaries into a bootable iso. Building `nano` or `Xmas` though isn't really clear because we don't have access to the branches when these were built.

> > Is there a particular place where we can communicate? The way to access current forms of communication is currently not really obvious or down. The forums have been down for a bit and I'm not too sure how to access the Freenode IRC channel.
>
>As to your question about communication, they have a subreddit (although with no more than 10 posts) [here](https://www.reddit.com/r/PureDarwin).

@ninjadev64 It's a little sad that it's really bare. I was hoping there would be a place that doubles as socializing and discussion about Darwin, since we're all here out of interest for the kernel, right? A telegram has been mentioned in #71 by @csekel, but there's no way of accessing it easily. The proposal for a Discord server seemed like a great idea.

I'll come back with more if I can get any of this stuff working on actual metal - or just a bump lol.

**Comment by pizazzo** [comment]
Another follow-up hoping that these questions could be answered (sorry if I'm getting annoying 😅) along with an update on my initial Linux/Windows build study.

I found out on #82 that `iig` - a part of the [DriverKit SDK](https://developer.apple.com/documentation/driverkit/creating_a_driver_using_the_driverkit_sdk) which compiles `*.iig` IOKit interface generator files - is required by PureDarwin to compile several files, and it's closed-source. A quick skim over the [darwin-sdk](https://github.com/darwin-on-arm/darwin-sdk) by the [Darwin on Arm ](https://github.com/darwin-on-arm/) people show that they don't really have it either. It's a non-trivial task, and will probably mean that a lot of the driver work will still have to be built on a macOS host at the very least.

_**Edit:** A search up on Darwin on Arm's [xnu](https://github.com/darwin-on-arm/xnu/) fork shows that they don't have any `*.iig` files. These were introduced as part of macOS 10.15 and DriverKit, so it makes sense that their fork doesn't have it. The XNU version in the PD trunk branch must be pretty recent?_

**Comment by nekename** [comment]
I'm going to set up an unofficial Discord server later today; I'll follow up with the link.

By the way @pizzabug, you posted your last comment twice.

**Comment by nekename** [comment]
If anyone cares: https://discord.gg/9kz8XXRRcT

**Comment by pizazzo** [comment]
Brief update. Firstly, I want to apologise; a lot of these questions are in places you can find - though, it'd probably help to put them in more obvious places. I'm writing up answers to my questions on the earlier comments so that people reading/watching this issue know I found answers, but I've updated my previous ones for easier reading.

> I'm not too sure if it can act as a drop-in replacement for the DarwinBuild toolchain, or replace Xcode requirements on DarwinBuild to allow it to run on Linux (#57); I'll still need to test this out. This isn't a 'must' since my situation is certainly an outlier, but it could help expand the number of developers that can work on the project.

It's not possible until an open-source alternative for `iig` is implemented. In terms of deploying to other platforms, someone could fork [Darwin on ARM](https://github.com/darwin-on-arm) on [darwin-sdk](https://github.com/darwin-on-arm/darwin-sdk) to expand their toolchain to fit PureDarwin's?

> A search up on Darwin on Arm's xnu fork shows that they don't have any *.iig files. These were introduced as part of macOS 10.15 and DriverKit, so it makes sense that their fork doesn't have it. The XNU version in the PD trunk branch must be pretty recent?

If the [homepage](http://www.puredarwin.org/) is still updated, we're at XNU 16.7(10.12.6). Build guide is [here](https://github.com/PureDarwin/Building-XNU-16.7-Kernel-Guide/).

> Is there a particular place where we can communicate? The way to access current forms of communication is currently not really obvious or down. The forums have been down for a bit and I'm not too sure how to access the Freenode IRC channel.

We have an unofficial discord server by @ninjadev64 (thank you!) but do we still need Freenode access to work with the development team?

Again I'll respond once I get it running on metal, or find something new. 😝

**Comment by nekename** [comment]
Just to let people know; the devs have allowed the previously unofficial server to become official (meaning yes, you can work with the dev team there)

**Comment by nekename** [comment]
Note: I found the Google Site versions of the Darwin 10 and Darwin 9 pages you linked in your first comment.
https://sites.google.com/a/puredarwin.org/puredarwin/developers/darwinbuild/troubleshooting/darwin-10
https://sites.google.com/a/puredarwin.org/puredarwin/developers/darwinbuild/troubleshooting/darwin-9
They are much cleaner, I assure you.

**Comment by csekel** [comment]
@ninjadev64 They may be cleaner though those pages and projects are out of date or no longer developed.

**Comment by nekename** [comment]
> @ninjadev64 They may be cleaner though those pages and projects are out of date or no longer developed. 

I doubt these specific pages on the new wiki have been updated, since they are a garbled unreadable mess?

**Comment by wjk** [comment]
Our code base is not a descendant of any of the old code bases you might find online, nor did we build the ISOs that are floating around the internet. We are not planning any architectural ports at this time, but any attempts at such would be welcome. That being said, there is already plenty to do in the current codebase. Port new Apple packages if you want, or feel free to make the existing code more readable! As far as `iig` goes, please see #92. Thanks everyone!

**Comment by beeselmane** [comment]
Hi, I was wondering something somewhat similar to what the others in this thread were asking, I've looked around the wiki/the rest of this org and on the web generally for info on this project, and it seems that save for @wjk's recent work on this repo/a bit of stuff on darwinbuild, no real work has been put in for a few years. Would anyone be interested in seeing if we can clean up some of the old info on the wiki and working to include Apple's new sources and writing patches to build PureDarwin with modern XNU and system utils (Apple's current Darwin is into major version 20)? I see some work seems to have been put in on this repo, but it doesn't seem quite right, with cmake issues and such.

I'm wondering which of the actual members of this org are still active? It seems @csekel has had some recent activity, as well as one of the other members of the org, but the other members have no commit history for at least the last year. Anyone looking to take another look at this project? I think the idea of an open source *nix system based on the XNU kernel and Apple's userland sources sounds pretty cool.

Is there any info on who the current team is for this project or who to contact with regard to developing it further?

**Comment by nekename** [comment]
Hey beeselmane, if you want you can join the Discord server where the developers (and the community) are willing to chat.

**Comment by wjk** [comment]
For future reference, darwinbuild is now deprecated. We worked with it for a long time, but I eventually found it too unsuitable for rapid iteration; it was also becoming less and less compatible with macOS as it Apple continued to lock it down. I have therefore marked its repo as archived.

**Comment by nekename** [comment]
@wjk You should be able to convert this issue to a discussion; please could you do that?

**Reply by wjk** [reply]
Your wish is my command!
